ARM, and Big endian testing

We don't do testing on PRs for Big endian or ARM CPUs.

There's some docker images which use qemu emulation, which (hopefully) make this fairly easy. See this comment: https://github.com/travis-ci/travis-ci/issues/8869#issuecomment-528986808

The travis ci s390x has been added now too. 

These ones are failing:

- [ ] aapolygon(surface, points, color): return None ... FAIL
- [ ] test_frombuffer_BGR (pygame.tests.image_test.ImageModuleTest) ... FAIL
- [ ] test_frombuffer_RGB (pygame.tests.image_test.ImageModuleTest) ... FAIL

```
FAIL: test_aapolygon (pygame.tests.gfxdraw_test.GfxdrawDefaultTest)
aapolygon(surface, points, color): return None
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.7.5/lib/python3.7/site-packages/pygame-2.0.0.dev11-py3.7-linux-s390x.egg/pygame/tests/gfxdraw_test.py", line 723, in test_aapolygon
    self.check_at(surf, posn, fg_adjusted)
  File "/home/travis/virtualenv/python3.7.5/lib/python3.7/site-packages/pygame-2.0.0.dev11-py3.7-linux-s390x.egg/pygame/tests/gfxdraw_test.py", line 57, in check_at
    self.assertEqual(sc, color, fail_msg)
AssertionError: (8, 64, 128, 255) != (128, 64, 8, 255) : (8, 64, 128, 255) != (128, 64, 8, 255) at (10, 15), bitsize: 24, flags: 0, masks: (255, 65280, 16711680, 0)
```


```
FAIL: test_frombuffer_BGR (pygame.tests.image_test.ImageModuleTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.7.5/lib/python3.7/site-packages/pygame-2.0.0.dev11-py3.7-linux-s390x.egg/pygame/tests/image_test.py", line 608, in test_frombuffer_BGR
    self.assertEqual(bgr_surf.get_at((0, 0)), pygame.Color(255, 10, 20))
AssertionError: (20, 10, 255, 255) != (255, 10, 20, 255)

======================================================================
FAIL: test_frombuffer_RGB (pygame.tests.image_test.ImageModuleTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.7.5/lib/python3.7/site-packages/pygame-2.0.0.dev11-py3.7-linux-s390x.egg/pygame/tests/image_test.py", line 581, in test_frombuffer_RGB
    self.assertEqual(rgb_surf.get_at((0, 0)), pygame.Color(255, 10, 20))
AssertionError: (20, 10, 255, 255) != (255, 10, 20, 255)
```

--- a/.github/workflows/build-debian-multiarch.yml
+++ b/.github/workflows/build-debian-multiarch.yml
@@ -1,0 +1,93 @@
+# Tests pygame on more exotic architectures. This is not something that is
+# actively supported, but source code support for this is nice to have. We
+# don't do any releases from here.
+
+name: Debian Multiarch
+
+# Run CI only when a release is created, on changes to main branch, or any PR 
+# to main. Do not run CI on any other branch. Also, skip any non-source changes 
+# from running on CI
+on:
+  release:
+    types: [created]
+  push:
+    branches: main
+    paths-ignore:
+      - 'docs/**'
+      - 'examples/**'
+      - '.gitignore'
+      - '*.rst'
+      - '*.md'
+      - '.github/workflows/*.yml'
+      # re-include current file to not be excluded
+      - '!.github/workflows/build-debian-multiarch.yml'
+
+  pull_request:
+    branches: main
+    paths-ignore:
+      - 'docs/**'
+      - 'examples/**'
+      - '.gitignore'
+      - '*.rst'
+      - '*.md'
+      - '.github/workflows/*.yml'
+      # re-include current file to not be excluded
+      - '!.github/workflows/build-debian-multiarch.yml'
+
+jobs:
+  build-multiarch:
+    name: Debian (Bullseye - 11) [${{ matrix.arch }}]
+    runs-on: ubuntu-22.04
+
+    strategy:
+      fail-fast: false  # if a particular matrix build fails, don't skip the rest
+      matrix:
+        # maybe more things could be added in here in the future (if needed)
+        arch: [s390x, ppc64le]
+
+    steps:
+    - uses: actions/checkout@v3.0.2
+
+    - name: Build sources and run tests
+      uses: uraimo/run-on-arch-action@v2.3.0
+      id: build
+      with:
+        arch: ${{ matrix.arch }}
+        distro: bullseye
+
+        # Not required, but speeds up builds
+        githubToken: ${{ github.token }}
+
+        # Create an artifacts directory
+        setup: mkdir -p ~/artifacts
+
+        # Mount the artifacts directory as /artifacts in the container
+        dockerRunArgs: --volume ~/artifacts:/artifacts
+
+        # The shell to run commands with in the container
+        shell: /bin/sh
+
+        # Install some dependencies in the container. This speeds up builds if
+        # you are also using githubToken. Any dependencies installed here will
+        # be part of the container image that gets cached, so subsequent
+        # builds don't have to re-install them. The image layer is cached
+        # publicly in your project's package repository, so it is vital that
+        # no secrets are present in the container state or logs.
+        install: |
+          apt-get update --fix-missing
+          apt-get upgrade -y
+          apt-get install libsdl2-dev libsdl2-image-dev libsdl2-mixer-dev libsdl2-ttf-dev libfreetype6-dev libportmidi-dev libjpeg-dev fontconfig -y
+          apt-get install python3-setuptools python3-dev python3-pip python3-wheel python3-sphinx -y
+
+        # Build a wheel, install it for running unit tests
+        run: |
+          export PIP_CONFIG_FILE=buildconfig/pip_config.ini
+          echo "\nBuilding pygame wheel\n"
+          python3 setup.py docs
+          pip3 wheel . --wheel-dir /artifacts -vvv
+          echo "\nInstalling wheel\n"
+          pip3 install --no-index --pre --find-links /artifacts pygame
+          echo "\nRunning tests\n"
+          export SDL_VIDEODRIVER=dummy
+          export SDL_AUDIODRIVER=disk
+          python3 -m pygame.tests -v --exclude opengl,music,timing --time_out 300

--- a/.github/workflows/build-ubuntu-sdist.yml
+++ b/.github/workflows/build-ubuntu-sdist.yml
@@ -54,10 +54,12 @@ jobs:
       run: |
         sudo apt-get update --fix-missing
         sudo apt-get upgrade
-        sudo apt-get install libsdl2-dev libsdl2-image-dev libsdl2-mixer-dev libsdl2-ttf-dev libfreetype6-dev libportmidi-dev libjpeg8-dev python3-setuptools python3-dev
+        sudo apt-get install libsdl2-dev libsdl2-image-dev libsdl2-mixer-dev libsdl2-ttf-dev libfreetype6-dev libportmidi-dev libjpeg-dev python3-setuptools python3-dev
         pip3 install sphinx numpy>=1.21.0
 
     - name: Make sdist and install it
+      env:
+        PIP_CONFIG_FILE: "buildconfig/pip_config.ini"
       run: |
         python3 setup.py docs
         python3 setup.py sdist
@@ -67,7 +69,7 @@ jobs:
       env:
         SDL_VIDEODRIVER: "dummy"
         SDL_AUDIODRIVER: "disk"
-      run: python3 -m pygame.tests -v --exclude opengl,timing --time_out 300
+      run: python3 -m pygame.tests -v --exclude opengl,music,timing --time_out 300
     
     - name: Test typestubs
       if: matrix.os == 'ubuntu-22.04' # run stubest only once

--- a/buildconfig/config_unix.py
+++ b/buildconfig/config_unix.py
@@ -8,12 +8,16 @@ from distutils.sysconfig import get_python_inc
 
 configcommand = os.environ.get('SDL_CONFIG', 'sdl-config',)
 configcommand = configcommand + ' --version --cflags --libs'
-localbase = os.environ.get('LOCALBASE', '')
+
 if os.environ.get('PYGAME_EXTRA_BASE', ''):
     extrabases = os.environ['PYGAME_EXTRA_BASE'].split(':')
 else:
     extrabases = []
 
+if os.environ.get('LOCALBASE', ''):
+    extrabases.append(os.environ['LOCALBASE'])
+
+extrabases.extend(("/usr", "/usr/local"))
 
 class DependencyProg:
     def __init__(self, name, envname, exename, minver, defaultlibs, version_flag="--version"):
@@ -135,10 +139,35 @@ def main(auto_config=False):
 
     #these get prefixes with '/usr' and '/usr/local' or the $LOCALBASE
     origincdirs = ['/include', '/include/SDL2']
-    origlibdirs = ['/lib', '/lib64', '/X11R6/lib',
-                   '/lib/i386-linux-gnu', '/lib/x86_64-linux-gnu',
-                   '/lib/arm-linux-gnueabihf/', '/lib/aarch64-linux-gnu/', 
-                   '/lib/powerpc64le-linux-gnu/']
+    origlibdirs = ['/lib', '/lib64', '/X11R6/lib']
+
+    # If we are on a debian based system, we also need to handle 
+    # /lib/<multiarch-tuple>
+    # We have a few commands to get the correct <multiarch-tuple>, we try those
+    # one by one till we get something that works
+    for cmd in (
+        "dpkg-architecture -qDEB_HOST_MULTIARCH",
+        "gcc -print-multiarch",
+        "gcc -dumpmachine",
+    ):
+        try:
+            f = os.popen(cmd)
+        except Exception:
+            # We don't bother here, instead we try the next fallback
+            continue
+
+        try:
+            stdout = f.read().strip()
+        finally:
+            if f.close() is not None:
+                # The command didn't exist successfully, the stdout we got is
+                # useless
+                stdout = ""
+
+        if stdout:
+            # found what we were looking for
+            origlibdirs.append(f"/lib/{stdout}")
+            break
 
     if 'ORIGLIBDIRS' in os.environ and os.environ['ORIGLIBDIRS'] != "":
         origlibdirs = os.environ['ORIGLIBDIRS'].split(":")
@@ -216,13 +245,6 @@ def main(auto_config=False):
     for extrabase in extrabases:
         incdirs += [extrabase + d for d in origincdirs]
         libdirs += [extrabase + d for d in origlibdirs]
-    incdirs += ["/usr"+d for d in origincdirs]
-    libdirs += ["/usr"+d for d in origlibdirs]
-    incdirs += ["/usr/local"+d for d in origincdirs]
-    libdirs += ["/usr/local"+d for d in origlibdirs]
-    if localbase:
-        incdirs = [localbase+d for d in origincdirs]
-        libdirs = [localbase+d for d in origlibdirs]
 
     for arg in DEPS[0].cflags.split():
         if arg[:2] == '-I':

--- a/src_c/SDL_gfx/SDL_gfxPrimitives.c
+++ b/src_c/SDL_gfx/SDL_gfxPrimitives.c
@@ -338,48 +338,33 @@ int _putPixelAlpha(SDL_Surface *dst, Sint16 x, Sint16 y, Uint32 color, Uint8 alp
 		break;
 
 	case 3: 
-		{		/* Slow 24-bpp mode, usually not used */
-			Uint8 Rshift8, Gshift8, Bshift8, Ashift8;
+		{		
+			/* Slow 24-bpp mode. This is patched on the pygame end because the
+			 * original code failed on big endian */
 			Uint8 *pixel = (Uint8 *) dst->pixels + y * dst->pitch + x * 3;
+			Uint8 *dR, *dG, *dB;
+			Uint8 sR, sG, sB;
 
-			Rshift = format->Rshift;
-			Gshift = format->Gshift;
-			Bshift = format->Bshift;
-			Ashift = format->Ashift;
+			SDL_GetRGB(color, format, &sR, &sG, &sB);
 
-			Rshift8 = Rshift / 8;
-			Gshift8 = Gshift / 8;
-			Bshift8 = Bshift / 8;
-			Ashift8 = Ashift / 8;
+#if (SDL_BYTEORDER == SDL_LIL_ENDIAN)
+            dR = pixel + (format->Rshift >> 3);
+            dG = pixel + (format->Gshift >> 3);
+            dB = pixel + (format->Bshift >> 3);
+#else
+            dR = pixel + 2 - (format->Rshift >> 3);
+            dG = pixel + 2 - (format->Gshift >> 3);
+            dB = pixel + 2 - (format->Bshift >> 3);
+#endif
 
 			if (alpha == 255) {
-				*(pixel + Rshift8) = color >> Rshift;
-				*(pixel + Gshift8) = color >> Gshift;
-				*(pixel + Bshift8) = color >> Bshift;
-				*(pixel + Ashift8) = color >> Ashift;
+				*dR = sR;
+				*dG = sG;
+				*dB = sB;
 			} else {
-				Uint8 dR, dG, dB, dA = 0;
-				Uint8 sR, sG, sB, sA = 0;
-
-				dR = *((pixel) + Rshift8);
-				dG = *((pixel) + Gshift8);
-				dB = *((pixel) + Bshift8);
-				dA = *((pixel) + Ashift8);
-
-				sR = (color >> Rshift) & 0xff;
-				sG = (color >> Gshift) & 0xff;
-				sB = (color >> Bshift) & 0xff;
-				sA = (color >> Ashift) & 0xff;
-
-				dR = dR + ((sR - dR) * alpha >> 8);
-				dG = dG + ((sG - dG) * alpha >> 8);
-				dB = dB + ((sB - dB) * alpha >> 8);
-				dA = dA + ((sA - dA) * alpha >> 8);
-
-				*((pixel) + Rshift8) = dR;
-				*((pixel) + Gshift8) = dG;
-				*((pixel) + Bshift8) = dB;
-				*((pixel) + Ashift8) = dA;
+				*dR = *dR + ((sR - *dR) * alpha >> 8);
+				*dG = *dG + ((sG - *dG) * alpha >> 8);
+				*dB = *dB + ((sB - *dB) * alpha >> 8);
 			}
 		}
 		break;
@@ -639,45 +624,32 @@ int _filledRectAlpha(SDL_Surface * dst, Sint16 x1, Sint16 y1, Sint16 x2, Sint16 
 		break;
 
 	case 3:
-		{			/* Slow 24-bpp mode, usually not used */
+		{	
+			/* Slow 24-bpp mode. This is patched on the pygame end because the
+			 * original code failed on big endian */
 			Uint8 *row, *pix;
-			Uint8 dR, dG, dB, dA;
-			Uint8 Rshift8, Gshift8, Bshift8, Ashift8;
+			Uint8 *dR, *dG, *dB;
+			Uint8 sR, sG, sB;
 
-			Rshift = format->Rshift;
-			Gshift = format->Gshift;
-			Bshift = format->Bshift;
-			Ashift = format->Ashift;
-
-			Rshift8 = Rshift / 8;
-			Gshift8 = Gshift / 8;
-			Bshift8 = Bshift / 8;
-			Ashift8 = Ashift / 8;
-
-			sR = (color >> Rshift) & 0xff;
-			sG = (color >> Gshift) & 0xff;
-			sB = (color >> Bshift) & 0xff;
-			sA = (color >> Ashift) & 0xff;
-
+			SDL_GetRGB(color, format, &sR, &sG, &sB);
 			for (y = y1; y <= y2; y++) {
 				row = (Uint8 *) dst->pixels + y * dst->pitch;
 				for (x = x1; x <= x2; x++) {
 					pix = row + x * 3;
 
-					dR = *((pix) + Rshift8);
-					dG = *((pix) + Gshift8);
-					dB = *((pix) + Bshift8);
-					dA = *((pix) + Ashift8);
+#if (SDL_BYTEORDER == SDL_LIL_ENDIAN)
+            		dR = pix + (format->Rshift >> 3);
+            		dG = pix + (format->Gshift >> 3);
+            		dB = pix + (format->Bshift >> 3);
+#else
+            		dR = pix + 2 - (format->Rshift >> 3);
+            		dG = pix + 2 - (format->Gshift >> 3);
+            		dB = pix + 2 - (format->Bshift >> 3);
+#endif
 
-					dR = dR + ((sR - dR) * alpha >> 8);
-					dG = dG + ((sG - dG) * alpha >> 8);
-					dB = dB + ((sB - dB) * alpha >> 8);
-					dA = dA + ((sA - dA) * alpha >> 8);
-
-					*((pix) + Rshift8) = dR;
-					*((pix) + Gshift8) = dG;
-					*((pix) + Bshift8) = dB;
-					*((pix) + Ashift8) = dA;
+					*dR = *dR + ((sR - *dR) * alpha >> 8);
+					*dG = *dG + ((sG - *dG) * alpha >> 8);
+					*dB = *dB + ((sB - *dB) * alpha >> 8);
 				}
 			}
 		}

--- a/src_c/mixer.c
+++ b/src_c/mixer.c
@@ -592,8 +592,10 @@ pg_mixer_get_init(PyObject *self, PyObject *_null)
         Py_RETURN_NONE;
 
     // create a signed or unsigned number of bits per sample
-    // XXX: When mixer is init'd with a format of -8, this returns +8
-    realform = (format & ~0xff) ? -(format & 0xff) : format & 0xff;
+    realform = SDL_AUDIO_BITSIZE(format);
+    if (SDL_AUDIO_ISSIGNED(format)) {
+        realform = -realform;
+    }
     return Py_BuildValue("(iii)", freq, realform, channels);
 }
 

--- a/test/image_test.py
+++ b/test/image_test.py
@@ -10,6 +10,13 @@ import pathlib
 from pygame.tests.test_utils import example_path, png, tostring
 import pygame, pygame.image, pygame.pkgdata
 
+sdl_image_svg_jpeg_save_bug = False
+_sdl_image_ver = pygame.image.get_sdl_image_version()
+if _sdl_image_ver is not None:
+    sdl_image_svg_jpeg_save_bug = (
+        _sdl_image_ver <= (2, 0, 5) and pygame.get_sdl_byteorder() == pygame.BIG_ENDIAN
+    )
+
 
 def test_magic(f, magic_hexes):
     """Tests a given file to see if the magic hex matches."""
@@ -92,6 +99,10 @@ class ImageModuleTest(unittest.TestCase):
                     img_file = io.BytesIO(img_bytes)
                     image = pygame.image.load(img_file)
 
+    @unittest.skipIf(
+        sdl_image_svg_jpeg_save_bug,
+        "SDL_image 2.0.5 and older has a big endian bug in jpeg saving",
+    )
     def testSaveJPG(self):
         """JPG equivalent to issue #211 - color channel swapping
 
@@ -1179,6 +1190,11 @@ class ImageModuleTest(unittest.TestCase):
         ]
 
         for filename, expected_color in filename_expected_color:
+            if filename.endswith("svg") and sdl_image_svg_jpeg_save_bug:
+                # SDL_image 2.0.5 and older has an svg loading bug on big
+                # endian platforms
+                continue
+
             with self.subTest(
                 f'Test loading a {filename.split(".")[-1]}',
                 filename="examples/data/" + filename,

